include/mesh.tcc

```diff
@@ -1644,13 +1648,19 @@ bool mpm::Mesh<Tdim>::generate_particles(const std::shared_ptr<mpm::IO>& io,
       auto particle_type =
           generator["particle_type"].template get<std::string>();
       // Material id
-      unsigned material_id = generator["material_id"].template get<unsigned>();
+      std::vector<unsigned> material_ids;
+      if (generator.at("material_id").is_array())
```

Comment:
If the input json is `[0, 1]` then we know it's an array. But if it is `[0]` would it it counted as array or scalar?

Reply:
array

Reply:
The input file needs to be `0` instead of `[0]` which is not quite intuitive right?

Reply:
You can have `0` too following the else clause if `is_array()` is false, this won't break the current `.json` file.

Reply:
But instead of having an `unsigned` you will have a vector of unsigned, which is not your design right?

Reply:
Even a single element like `[0]` should behave like an array.

Reply:
> But instead of having an `unsigned` you will have a vector of unsigned, which is not your design right?

That's the proposed implementation. User can either use a scalar `0` (the current interface) or an array `[0, 1]` as an input. The latter one is needed for two-phase or three-phase particles.

Reply:
Thanks @bodhinandach , yes, so keeping scalar ensures backward compatibility